qt/widgets/instrumentview/src/InstrumentWidgetDecoder.cpp

```diff
@@ -0,0 +1,321 @@
+  surface->setPeakLabelPrecision(map[QString("labelPrecision")].toInt());
+  surface->setShowPeakRelativeIntensityFlag(
+      map[QString("showRelativeIntensity")].toBool());
+
+  // Load color bar
+  this->decodeColorBar(map[QString("colorBar")].toMap(), obj->m_colorBarWidget);
+}
+
+void InstrumentWidgetDecoder::decodeColorBar(const QMap<QString, QVariant> &map,
+                                             ColorBar *bar) {
+  const auto scaleType = map[QString("scaleType")].toInt();
+  const auto power = map[QString("power")].toString().toDouble();
+  const auto min = map[QString("min")].toString().toDouble();
+  const auto max = map[QString("max")].toString().toDouble();
+
+  bar->setScaleType(scaleType);
+  try {
+    bar->setNthPower(power);
+  } catch (const std::runtime_error &) {
+    // Do nothing, because this is where the power was loaded in as 0.
+  }
+
+  bar->setMinValue(min);
+  bar->setMaxValue(max);
+}
+
+void InstrumentWidgetDecoder::decodeTreeTab(const QMap<QString, QVariant> &map,
+                                            InstrumentWidgetTreeTab *obj) {
+  auto names = map[QString("expandedItems")].toList();
+  if (names.size() > 0) {
+    for (auto &name : names) {
+      auto index = obj->m_instrumentTree->findComponentByName(name.toString());
+      obj->m_instrumentTree->setExpanded(index, true);
+    }
+    obj->selectComponentByName(map[QString("selectedComponent")].toString());
+  }
+}
+
+void InstrumentWidgetDecoder::decodePickTab(const QMap<QString, QVariant> &map,
+                                            InstrumentWidgetPickTab *obj) {
+  obj->m_zoom->setChecked(map[QString("zoom")].toBool());
+  obj->m_edit->setChecked(map[QString("edit")].toBool());
+  obj->m_ellipse->setChecked(map[QString("ellipse")].toBool());
+  obj->m_rectangle->setChecked(map[QString("rectangle")].toBool());
+  obj->m_ring_ellipse->setChecked(map[QString("ringEllipse")].toBool());
+  obj->m_ring_rectangle->setChecked(map[QString("ringRectangle")].toBool());
+  obj->m_free_draw->setChecked(map[QString("freeDraw")].toBool());
+  obj->m_one->setChecked(map[QString("one")].toBool());
+  obj->m_tube->setChecked(map[QString("tube")].toBool());
+  obj->m_peak->setChecked(map[QString("peak")].toBool());
+  obj->m_peakSelect->setChecked(map[QString("peakSelect")].toBool());
+}
+
+void InstrumentWidgetDecoder::decodeActor(
+    const QMap<QString, QVariant> &map, std::unique_ptr<InstrumentActor> &obj) {
+  obj->loadColorMap(map[QString("fileName")].toString());
+
+  this->decodeBinMasks(map[QString("binMasks")].toList(), obj->m_maskBinsData);
+}
+
+void InstrumentWidgetDecoder::decodeBinMasks(const QList<QVariant> &list,
+                                             MaskBinsData &obj) {
+  for (const auto &item : list) {
+    const auto itemMap = item.toMap();
+    const auto range = itemMap[QString("range")].toList();
+    double start = range[0].toDouble();
+    double end = range[1].toDouble();
```

Comment:
Again with what I said previously, about this being a coder defined constraint of requiring a QList of length 2 to be made because that is what is defined.